src/hooks/useAudioPlayer.ts

```diff
@@ -2,221 +2,309 @@ import { Callbacks, CogsClientMessage } from '@clockworkdog/cogs-client';
+      setAudioClipPlayers((previousClipPlayers) => {
+        const clips = { ...previousClipPlayers };
+        let clip = clips[clipPath];
         if (!clip) {
-          clip = createClip(clipId, { preload: false, ephemeral: true });
-          clips[clipId] = clip;
-        }
-
-        // Play the clip first to get a new ID for use with subsequent options
-        const id = clip.player.play();
-        clip.player.loop(loop, id);
-
-        // Start fade when clip starts
-        if (typeof fade === 'number' && !isNaN(fade) && fade > 0) {
-          clip.player.volume(0, id);
-          clip.player.once(
-            'play',
-            () => {
-              clip.player.once(
-                'fade',
-                () => {
-                  clip.player.volume(volume, id);
-                },
-                id
-              );
-              clip.player.fade(0, volume, fade * 1000, id);
-            },
-            id
-          );
+          clip = createClip(clipPath, { preload: false, ephemeral: true });
+          clips[clipPath] = clip;
         } else {
-          clip.player.volume(volume, id);
+          clip = { ...clip };
+          clips[clipPath] = clip;
         }
+
+        const pausedSoundIds = Object.entries(clip.activeClips)
+          .filter(([, { state }]) => state === ActiveAudioClipState.Paused || state === ActiveAudioClipState.Pausing)
+          .map(([id]) => parseInt(id));
+
+        // Paused clips need to be played again
+        pausedSoundIds.forEach((soundId) => {
+          clip.player.play(soundId);
+        });
+
+        // If no currently paused/pausing clips, play a new clip
+        const newSoundIds =
+          pausedSoundIds.length > 0
+            ? []
+            : [
+                (() => {
+                  const soundId = clip.player.play();
+                  return soundId;
+                })(),
+              ];
+
+        [...pausedSoundIds, ...newSoundIds].forEach((soundId) => {
+          // Cleanup any old callbacks first
+          clip.player.off('fade', undefined, soundId);
+          clip.player.off('end', undefined, soundId);
+          clip.player.off('stop', undefined, soundId);
+          clip.player.loop(loop, soundId);
+
+          clip.player.once('stop', () => handleStoppedClip(clipPath, soundId), soundId);
+
+          // Looping clips fire the 'end' callback on every loop
+          if (!loop) {
+            clip.player.once('end', () => handleStoppedClip(clipPath, soundId), soundId);
+          }
+
+          const activeClip: ActiveClip = {
+            state: ActiveAudioClipState.Playing,
+            loop,
+            volume,
+          };
+
+          // Start fade when clip starts
+          if (isFadeValid(fade)) {
+            clip.player.volume(0, soundId);
+            clip.player.once(
+              'play',
+              () => {
+                // clip.player.once('fade', () => clip.player.volume(volume, soundId), soundId);
```

Comment:
Why the change from `once` here?

Reply:
This was redundant, the `fade` does indeed change the internal Howl player's volume. I'll remove the line.